### ISSUE TITLE
avoid Cython dependency for general users

### ISSUE THREAD

**Issue by kmaehashi** [jubatus/embedded-jubatus-python#18]
Cython docs says:

http://cython.readthedocs.io/en/latest/src/reference/compilation.html#distributing-cython-modules

> It is strongly recommended that you distribute the generated .c files as well as your Cython sources, so that users can install your module without needing to have Cython available.

So I think it's better to include ``embedded_jubatus.cpp`` as a part of distribution, so that users don't have to install Cython.

**Comment by kazuki** [comment]
:+1: I missed that note.

Another approach is to register manylinux1 binary wheel to pypi.
But this approach requires static link to jubatus_core/msgpack.

**Comment by kazuki** [comment]
I tests pip wheel & auditwheel. embedded-jubatus depends many many libraries....
Because, jubatus (not jubatus-core) depends zookeeper, apr, ...
If distribute manylinux1 wheel package, we need to avoid jubatus dependency.

```
Archive:  wheelhouse/embedded_jubatus-1.0.0-cp35-cp35m-manylinux1_x86_64.whl                                                
 Length   Method    Size  Cmpr    Date    Time   CRC-32   Name
--------  ------  ------- ---- ---------- ----- --------  ----
 9936328  Defl:N  2296637  77% 11-08-2016 16:29 7aa61c1a  embedded_jubatus.cpython-35m-x86_64-linux-gnu.so
  174656  Defl:N    61543  65% 11-08-2016 16:29 6eb049fb  .libsembedded_jubatus/libexpat-1c42acd1.so.1.6.0
   38336  Defl:N     8034  79% 11-08-2016 16:29 f710d3f9  .libsembedded_jubatus/libjubatus_util_concurrent-2ae3830e.so.1.0.0
   42616  Defl:N    11305  74% 11-08-2016 16:29 2b174e4f  .libsembedded_jubatus/libjubatus_util_system-e41b6aef.so.1.0.0
    4696  Defl:N     1290  73% 11-08-2016 16:29 5ddc9103  .libsembedded_jubatus/libjubatus_util_lang-b8fd68ad.so.1.0.0
 1604088  Defl:N   709675  56% 11-08-2016 16:29 88710324  .libsembedded_jubatus/libdb-4-ae5dba0d.8.so
 4483552  Defl:N   555907  88% 11-08-2016 16:29 3c8258da  .libsembedded_jubatus/liblog4cxx-6ff44b42.so.10.0.0
 1102840  Defl:N   129993  88% 11-08-2016 16:29 52d5b4b8  .libsembedded_jubatus/libjubaserv_common-5bf0af04.so.1.0.0
  193440  Defl:N    72334  63% 11-08-2016 16:29 916ea3a3  .libsembedded_jubatus/libaprutil-1-efc36894.so.0.5.4
   21728  Defl:N     7372  66% 11-08-2016 16:29 0078afad  .libsembedded_jubatus/libuuid-9c9c4ab4.so.1.3.0
  108448  Defl:N    10178  91% 11-08-2016 16:29 5bdab385  .libsembedded_jubatus/libjubaserv_common_logger-9d58e9c8.so.1.0.0
  246160  Defl:N    95685  61% 11-08-2016 16:29 c711558f  .libsembedded_jubatus/libapr-1-0dc794d3.so.0.5.2
  147544  Defl:N    47007  68% 11-08-2016 16:29 636fbda0  .libsembedded_jubatus/libjubatus_util_data-8121386f.so.1.0.0
   30024  Defl:N     1740  94% 11-08-2016 16:29 8ca43e1f  .libsembedded_jubatus/libjubatus_util-22a5e766.so.1.0.0
11776696  Defl:N  1416583  88% 11-08-2016 16:29 d7d03bc9  .libsembedded_jubatus/libjubatus_core-f26a12a1.so.1.0.0
  435560  Defl:N   111921  74% 11-08-2016 16:29 f3670e69  .libsembedded_jubatus/libonig-999b4ddf.so.2.0.0
    9168  Defl:N     1966  79% 11-08-2016 16:29 4fa3d0ed  .libsembedded_jubatus/libjubatus_util_math-652ac43f.so.1.0.0
  114872  Defl:N    41349  64% 11-08-2016 16:29 5f299b89  .libsembedded_jubatus/libzookeeper_mt-ae4e0180.so.2.0.0
   51128  Defl:N    22133  57% 11-08-2016 16:29 53fc5b8e  .libsembedded_jubatus/libgdbm-1aa6b078.so.4.0.0
   26328  Defl:N    10043  62% 11-08-2016 16:29 510ecc03  .libsembedded_jubatus/libmsgpack-95a9a89a.so.3.0.0
  211344  Defl:N    52013  75% 11-08-2016 16:29 b64cc11e  .libsembedded_jubatus/libjubatus_util_text-5d2944cf.so.1.0.0
    2941  Defl:N     1504  49% 11-08-2016 16:29 ac6f0fc1  embedded_jubatus-1.0.0.dist-info/RECORD
    2267  Defl:N     1011  55% 11-08-2016 07:27 98d87e3e  embedded_jubatus-1.0.0.dist-info/METADATA
     109  Defl:N      106   3% 11-08-2016 16:29 88553c83  embedded_jubatus-1.0.0.dist-info/WHEEL
      17  Defl:N       19 -12% 11-08-2016 07:27 61a68ab7  embedded_jubatus-1.0.0.dist-info/top_level.txt
     915  Defl:N      491  46% 11-08-2016 07:27 90158ebf  embedded_jubatus-1.0.0.dist-info/metadata.json
    1533  Defl:N      715  53% 11-08-2016 07:27 b1bbcaf4  embedded_jubatus-1.0.0.dist-info/DESCRIPTION.rst
--------          -------  ---                            -------
30767334          5668554  82%                            27 files
```

**Comment by kmaehashi** [comment]
In the previous release we added pure-python model manipulation feature (including CRC32 validations etc.) to jubakit:

* https://github.com/jubatus/jubakit/blob/master/jubakit/model.py#L53
* http://jubat.us/en/jubakit/api/jubakit.html#module-jubakit.model
* http://jubat.us/en/jubakit/guide/jubamodel.html

So using it instead of jubatus C++ library may be a possible solution.

**Comment by kmaehashi** [comment]
... or, just copying ``jubatus::server::common::calc_crc32`` function into embedded_jubatus source tree may be enough.

**Comment by kazuki** [comment]
after #19, manylinux1 package contains files below:

```
Archive:  embedded_jubatus-1.0.0.post1-cp35-cp35m-manylinux1_x86_64.whl
 Length   Method    Size  Cmpr    Date    Time   CRC-32   Name
--------  ------  ------- ---- ---------- ----- --------  ----
 9788720  Defl:N  2296929  77% 11-08-2016 19:09 c79916cd  embedded_jubatus.cpython-35m-x86_64-linux-gnu.so
   26328  Defl:N    10043  62% 11-08-2016 19:09 510ecc03  .libsembedded_jubatus/libmsgpack-95a9a89a.so.3.0.0
    4696  Defl:N     1290  73% 11-08-2016 19:09 5ddc9103  .libsembedded_jubatus/libjubatus_util_lang-b8fd68ad.so.1.0.0
  211344  Defl:N    52013  75% 11-08-2016 19:09 b64cc11e  .libsembedded_jubatus/libjubatus_util_text-5d2944cf.so.1.0.0
11776696  Defl:N  1416583  88% 11-08-2016 19:09 d7d03bc9  .libsembedded_jubatus/libjubatus_core-f26a12a1.so.1.0.0
  435560  Defl:N   111921  74% 11-08-2016 19:09 f3670e69  .libsembedded_jubatus/libonig-999b4ddf.so.2.0.0
    9168  Defl:N     1966  79% 11-08-2016 19:09 4fa3d0ed  .libsembedded_jubatus/libjubatus_util_math-652ac43f.so.1.0.0
   30024  Defl:N     1740  94% 11-08-2016 19:09 8ca43e1f  .libsembedded_jubatus/libjubatus_util-22a5e766.so.1.0.0
   38336  Defl:N     8034  79% 11-08-2016 19:09 f710d3f9  .libsembedded_jubatus/libjubatus_util_concurrent-2ae3830e.so.1.0.0
   42616  Defl:N    11305  74% 11-08-2016 19:09 2b174e4f  .libsembedded_jubatus/libjubatus_util_system-e41b6aef.so.1.0.0
  147544  Defl:N    47007  68% 11-08-2016 19:09 636fbda0  .libsembedded_jubatus/libjubatus_util_data-8121386f.so.1.0.0
    1862  Defl:N      944  49% 11-08-2016 19:09 341dca15  embedded_jubatus-1.0.0.post1.dist-info/RECORD
    2269  Defl:N     1013  55% 11-08-2016 10:08 0d9acd5b  embedded_jubatus-1.0.0.post1.dist-info/METADATA
     109  Defl:N      106   3% 11-08-2016 19:09 88553c83  embedded_jubatus-1.0.0.post1.dist-info/WHEEL
      17  Defl:N       19 -12% 11-08-2016 10:08 61a68ab7  embedded_jubatus-1.0.0.post1.dist-info/top_level.txt
     917  Defl:N      493  46% 11-08-2016 10:08 8061b1e9  embedded_jubatus-1.0.0.post1.dist-info/metadata.json
    1533  Defl:N      715  53% 11-08-2016 10:08 b1bbcaf4  embedded_jubatus-1.0.0.post1.dist-info/DESCRIPTION.rst
--------          -------  ---                            -------
22517739          3962121  82%                            17 files

```

**Comment by kmaehashi** [comment]
👍

**Comment by kazuki** [comment]
Currently, setup.py detects numpy installation to build fit/predict method.
And cython generates numpy depended `embedded_jubatus.cpp`.

I think there are three options to include `embedded_jubatus.cpp`.

1. Mandatory to depend on numpy
2. Includes two version of `embedded_jubatus.cpp`.
  * numpy version (enables fit/predict method)
  * no numpy version (disable fit/predict method)
3. Avoid numpy dependency
  * Avoid typed variable definition (Uses python dynamic typing)
  * Maybe, fit/predict method performance is slower
  * It is not certain whether it is possible method

I will test method 3 in next week.

**Comment by kazuki** [comment]
[PEP571 manylinux2010](https://www.python.org/dev/peps/pep-0571/)  released.

I pushed #56 to fix this issue.

**Comment by shiodat** [comment]
This issue fixed via #56, thanks!
